docs/adr/0008-redis-lua-scripts.md

```diff
@@ -0,0 +1,69 @@
+# Redis Lua Scripts
+
+- Status: draft
+- Deciders: dannycoates + TBD
+- Date: 2019-11-15
+
+## Context and Problem Statement
+
+We're planning on expanding our use of Redis to replace some or all of our MySQL database for OAuth. Our `fxa-shared/redis` module also requires maintenance as part of our modernization effort. The expansion will include use cases that require new indexing, join-like, and transactional operations.
+
+## Decision Drivers
+
+- Need for robust redis transactions
+- Increased use of redis for oauth
+- Updating module dependencies
+
+## Considered Options
+
+- Continue expanding on fxa-shared/redis
+- Use lua scripts
+
+## Decision Outcome
+
+TBD
+
+### Positive Consequences
+
+- TBD
+
+### Negative Consequences
+
+- TBD
+
+## Pros and Cons of the Options
+
+### Continue expanding on fxa-shared/redis
+
+`fxa-shared/redis` uses a connection pool and utilizes the bluebird module disposer feature for transaction isolation and connection management. It currently depends on an older redis module that hasn't been updated in two years. With this option we'll likely switch to ioredis but maintain our connection pool. We'll also need to add new functionality to handle new transactions that the existing `update` function can't handle. We'll need to design a retry mechanism to handle write conflicts because the existing code simply ignores these, which is ok for its current use case, but insufficient for some new cases. Transactional cases will require more code and error handling using this method.
```

Comment:
From https://github.com/mozilla/fxa/issues/2999, when an error is thrown in redis they don't bubble up the stacktrace nicely. Do lau scripts provide additional debug information or are they easier to debug?

Reply:
Good question. I know that lua scripts are no _worse_ for debugging. While working on the prototype I was able to see stack traces down to the js function that made the call to redis, like `pruneSessionTokens` in the logs. The error _message_ from lua is there too, but not a stacktrace.

For debugging individual lua scripts during dev [redis-cli --ldb](https://redis.io/topics/ldb) worked well.

Reply:
I actually wonder if eliminating the indirection of the connection pool and promisified wrapper might fix stack traces.

Reply:
There's `verror` in use now as well which should help with the stack traces.